keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,886 @@
+operations will get retried and might eventually succeed, for example
+because storage capacity gets freed up or extended. What does not
+happen is that the pod is re-scheduled to some other node which has
+enough storage capacity.
+
+## Motivation
+
+### Goals
+
+The goal of this KEP is to increase the chance of choosing a node for
+which volume creation will succeed by tracking the currently available
+capacity available through a CSI driver and using that information
+during pod scheduling.
+
+Although the issue is more common with storage that is local to a node
+(LVM, PMEM), it may also occur for storage providers that have
+capacity limits for other topology segments (rack, data center,
+region, etc.). Capacity tracking is meant to be generic enough to
+support all of these cases.
+
+Inline volumes currently do not have a standardized way of specify the
+size; this KEP will introduce such a field and ensure that capacity
+tracking also works for drivers which only support ephemeral inline
+volumes.
+
+For persistent volumes, capacity will be tracked per storage class, so
+additional parameters in a storage class are taken into account. This
+is important because those parameters might have a significant impact on
+how much space a new volume actually needs in the underlying storage
+system (for example, an LVM volume with mirroring needs more space
+than a LVM volume that is striped). For ephemeral volumes there is no
+storage class, so tracking is only done per node and provisioner.
+
+### Non-Goals
+
+Only CSI drivers will be supported.
+
+The Kubernetes scheduler could try to anticipate the effect of
+creating multiple volumes concurrently. But this depends on knowledge
+about internal driver details that Kubernetes doesn’t have, so pending
+volume operations are simply ignored when making scheduling decisions.
+
+Because of that and also for other reasons (capacity changed via
+operations outside of Kubernetes, like creating or deleting volumes,
+or expanding the storage), it is expected that pod scheduling may
+still end up with a node from time to time where volume creation then
+fails. Rolling back in this case is complicated and outside of the
+scope of this KEP. For example, a pod might use two persistent
+volumes, of which one was created and the other not, and then it
+wouldn’t be obvious whether the existing volume can or should be
+deleted.
+
+For persistent volumes that get created independently of a pod nothing
+changes: it’s still the responsibility of the CSI driver to decide how
+to create the volume and then communicate back through topology
+information where pods using that volume need to run.
+
+Inline volumes could be extended to reference a storage class, which
+then could be used to handle more complex situations (like the LVM
+mirror vs. striped case) also for inline volumes. But this is outside
+the scope of this KEP.
+
+## Proposal
+
+### User Stories
```

Comment:
BTW, I find it odd that the KEP template has "user stories" under "proposal". A much better place would be in the previous section, "motivation". Shall I move it?

Reply:
Yes, that makes sense: bbe036bf0f61882a94e13e05a12e6ceaefc8717d.

Let me punt on the diagram for now, though. If this KEP gets rejected, then drawing diagrams would just be a waste of time. Unless it gets rejected because there is no diagram... :man_shrugging: